### ISSUE TITLE
Immutable X - NFT Bridge between L1 and StarkNet

### ISSUE THREAD

**Issue by rzmahmood** [immutable/imx-starknet#3]
# Immutable X - NFT Bridge between L1 and StarkNet

This issue serves as a place to aggregate and discuss feedback on an NFT Bridge Design on StarkNet.

Immutable X is working on an NFT Bridge which allows for projects to move from L1 Ethereum to L2 StarkNet. We believe that such a bridge can enable existing projects on L1 to utilize the benefits of StarkNet.

##  Design
The current high-level design is as such. 

On L1, there exist two contracts - a bridge contract and a registry. The registry serves as the source of truth for mappings of L1 contract addresses to L2 contract addresses. New pairs of addresses can be added by an authorized entity, in this case, Immutable X. The expectation is that project will reach out to Immutable X with their L1 NFT and L2 NFT addresses, Immutable X will validate it, then add it to the registry.

The L1 bridge contract uses this registry for L1 to L2 mappings. If a mapping on the registry exists for an NFT project, then owners of the token for the project can deposit one or more tokens on the L1 Bridging contract. When depositing, token owners will be required to provide their StarkNet address. The given address will receive the tokens on L2. Depositing will lock up the tokens in an Escrow contract and send a message to the L2 Bridge with a payload containing the token address, token ids, and L2 address of the new owner. The message to L2 will attempt to invoke the `handle_deposit` function on L2.

![image](https://user-images.githubusercontent.com/35128199/170401314-4243d562-4fe8-4b5e-94d4-da892dc243cb.png)

Once a message has been received on L2, basic validation will be done to ensure a valid source, then a mapping of the L2 NFT address to the L1 NFT address will be stored on the L2 Bridge if it doesn't exist already. The L2 Bridge contract will then mint the tokens on the L2 NFT contract, which will mint NFTs to the appropriate owner. The expectation is that the L2 Bridge contract will have permission to mint. In the off chance that the process fails, for example, due to a faulty L2 NFT contract or invalid permissions, a recovery mechanism exists.

We expect that the recovery mechanism will not be necessary unless due to a faulty L2 NFT contract. In this case, we use [the message cancellation flow by Starkware](https://community.starknet.io/t/l1-to-l2-message-cancellation/212) which allows a user to withdraw their NFT on L1 after a 5 day wait period in the scenario that the NFT failed to mint on L2.

For the withdrawal to L1, token holders will interact with the L2 Bridge and will need to provide their L1 recipient address. The L2 bridge must also have permission to execute `permissionedBurn` on the L2 NFT contract. The L2 bridge will recursively burn the tokens on L2, then send a message to L1, allowing the recipient to claim their NFTs on L1 when they are ready. Currently, only tokens that have been bridged from L1 to L2 can be withdrawn back to L1, however, with community feedback and discussion, our design could support this.

![image](https://user-images.githubusercontent.com/35128199/170401349-c3c2480f-288c-45ee-8a26-28fd4ef2451e.png)

##  Reasoning

We explored the option of having a decentralized L1 registry, where rather than only allowing Immutable X to update the registry, we allowed the L1 contract owner to update. This would involve checking the result of `owner()` on the NFT contract during the registry update process. This is problematic as many contracts are not Ownable or use a multi-sig. Furthermore, there is no way to atomically validate that the L2 NFT contract owner is the same as the L1 NFT contract owner. To validate this, it would involve sending a message to L2, getting authorization from the L2 NFT contract owner, then sending a message back to L2. This is non-trivial and introduces a significant UX barrier. The danger of not validating the project owner is that an arbitrary user could register a malicious NFT contract on L2, for example, one that blocks transfers or worse. We explored other projects, including NFT bridges on optimism, and found that a centralized registry process is acceptable.

We also explored the idea of having a Standard NFT contract on L2, which would be the default contract that must be used when bridging to L2. This would prevent the malicious user scenario and would mean that an explicit registration may not be required but project owners would have no control over their L2 NFT contract. Furthermore, there is no way to deploy a contract from a contract on StarkNet and would require off-chain and centralized automation.

## TLDR

We have an NFT bridge design that aims to address many common issues with NFT bridging. We're interested in how your project could use an NFT bridge and how we can better improve the design to serve the community. Here are some questions to get us started:

- Does your project minted on StarkNet need the option to Bridge to L1?
- How could we make the process of being added to the registry easier for you?

**Comment by marcellobardus** [comment]
That's very exciting!
I think that the design is definitely valid and correct however I think that Starknet's cheap computation can be leveraged by incorporating a solution like [Fossil](https://github.com/OilerNetwork/fossil) even more.
In a nutshell Fossil allows to prove any claims against the L1 block's state, receipts and transaction roots.
Each message sent from L1 to L2 costs approximately `72k gas`, that design requires sending such message every time an NFT is bridged.
Assuming that 100 NFTs are transferred between blocks `x` and `z` on L1, 100 messages are required.
On the other hand with Fossil only one message is required, an L1 blockhash needs to be sent as a message to L2.
Such hash is later used to reconstruct the block header and access the required roots.
By trading of the latency all these 100 NFTs can be claimed on L2 by sending only 1 message.
The trade-off comes from the fact that whoever requested it's NFT to be sent at time `x`, his request won't be processable unless `blockhash(x)` or higher is sent to L2.
Which is not a big issue, as the sequencer has a short delay in picking up messages anyways.
Furthermore Fossil allows to prove not only state but also logs.
That said a design where the user proves to an L2 smart contract that a transaction resulting in a log `ERC721.Transfer(user, escrow, tokenId)` was included in a block between `x` and `z` becomes possible.
Combining it all together a design implementing Fossil might have the following properties.
* The number of L1 -> L2 messages can be traded off for latency, L1 costs can be mutualised between users.
* No approvals are needed, on L1 the user only needs to transfer the NFT from his wallet to the escrow.
* The `eth -> starknet` users registry is no longer required because it can be replaced with a `ecrecover` operation on Starknet.
In the next Cairo and Starknet releases primitives like `keccak256` and `ecrecover` might be turned into builtins, lowering the cost of using such design to a negligible amount.

**Comment by Orland0x** [comment]
> That's very exciting! I think that the design is definitely valid and correct however I think that Starknet's cheap computation can be leveraged by incorporating a solution like [Fossil](https://github.com/OilerNetwork/fossil) even more. In a nutshell Fossil allows to prove any claims against the L1 block's state, receipts and transaction roots. Each message sent from L1 to L2 costs approximately `72k gas`, that design requires sending such message every time an NFT is bridged. Assuming that 100 NFTs are transferred between blocks `x` and `z` on L1, 100 messages are required. On the other hand with Fossil only one message is required, an L1 blockhash needs to be sent as a message to L2. Such hash is later used to reconstruct the block header and access the required roots. By trading of the latency all these 100 NFTs can be claimed on L2 by sending only 1 message. The trade-off comes from the fact that whoever requested it's NFT to be sent at time `x`, his request won't be processable unless `blockhash(x)` or higher is sent to L2. Which is not a big issue, as the sequencer has a short delay in picking up messages anyways. Furthermore Fossil allows to prove not only state but also logs. That said a design where the user proves to an L2 smart contract that a transaction resulting in a log `ERC721.Transfer(user, escrow, tokenId)` was included in a block between `x` and `z` becomes possible. Combining it all together a design implementing Fossil might have the following properties.
> 
> * The number of L1 -> L2 messages can be traded off for latency, L1 costs can be mutualised between users.
> * No approvals are needed, on L1 the user only needs to transfer the NFT from his wallet to the escrow.
> * The `eth -> starknet` users registry is no longer required because it can be replaced with a `ecrecover` operation on Starknet.
>   In the next Cairo and Starknet releases primitives like `keccak256` and `ecrecover` might be turned into builtins, lowering the cost of using such design to a negligible amount.

Love this! Just wanted to add that gas costs can be further mutualised as multiple applications that use Fossil can share the same block hash registry and hence share the costs of sending over the block hashes. Snapshot X being one of them.

**Comment by zetsuboii** [comment]
Have you checked Mostar? My design was very similar to this
https://github.com/zetsuboii/cairo-projects/tree/mostar/mostar

**Comment by rzmahmood** [comment]
> That's very exciting! I think that the design is definitely valid and correct however I think that Starknet's cheap computation can be leveraged by incorporating a solution like [Fossil](https://github.com/OilerNetwork/fossil) even more. In a nutshell Fossil allows to prove any claims against the L1 block's state, receipts and transaction roots. Each message sent from L1 to L2 costs approximately `72k gas`, that design requires sending such message every time an NFT is bridged. Assuming that 100 NFTs are transferred between blocks `x` and `z` on L1, 100 messages are required. On the other hand with Fossil only one message is required, an L1 blockhash needs to be sent as a message to L2. Such hash is later used to reconstruct the block header and access the required roots. By trading of the latency all these 100 NFTs can be claimed on L2 by sending only 1 message. The trade-off comes from the fact that whoever requested it's NFT to be sent at time `x`, his request won't be processable unless `blockhash(x)` or higher is sent to L2. Which is not a big issue, as the sequencer has a short delay in picking up messages anyways. Furthermore Fossil allows to prove not only state but also logs. That said a design where the user proves to an L2 smart contract that a transaction resulting in a log `ERC721.Transfer(user, escrow, tokenId)` was included in a block between `x` and `z` becomes possible. Combining it all together a design implementing Fossil might have the following properties.
> 
> * The number of L1 -> L2 messages can be traded off for latency, L1 costs can be mutualised between users.
> * No approvals are needed, on L1 the user only needs to transfer the NFT from his wallet to the escrow.
> * The `eth -> starknet` users registry is no longer required because it can be replaced with a `ecrecover` operation on Starknet.
>   In the next Cairo and Starknet releases primitives like `keccak256` and `ecrecover` might be turned into builtins, lowering the cost of using such design to a negligible amount.

@marcellobardus 
Have reached out to do a synced session to discussion for this solution. I'll summarise the meeting notes here afterward.. I've briefly explored Fossil and here are a few key discussion points:

- Our current design means that for successful flows, users do not need to do an L2 transaction. Using Fossil would require that on L2, someone provides the proof and the full path for each block header. How do we access the full path and proof and who is responsible for serving this data?
- It seems that with Fossil, that each block header is a new message to L2. This would mean we only save gas costs if there are multiple NFTs being bridged in a single block. If there was only 1 NFT being deposited per block, the gas costs would be similar. Is my understanding correct?
- I see that Fossil has APIs. Is this solution exposed to downtime in the Fossil APIs? Is there a business model for Fossil?
- Is there any consideration for the withdraw flow? What if someone withdraws back to L1, then uses an old block hash proof on L2 again, are we protected from such replay attacks?
- You mention that we no longer need a user registry. Our design doesn't have a user registry. It has a registry of mappings between L1 NFT contract address and L2 NFT contract address. ecrecover wouldn't be useful here as contracts can't directly initiate arbitrary transactions like an EOA. 

Keen to understand this more as I do believe the idea of using L1 state roots and proving on L2 is valuable.

> Have you checked Mostar? My design was very similar to this https://github.com/zetsuboii/cairo-projects/tree/mostar/mostar

@zetsuboii 
I've seen your design. A few points on your design:
- How do you account for liquidity fragmentation on L2. People could deploy multiple mostar instances for the same project and it would have multiple L2 Token contracts representing the same L1 Token. Our design aims to keep the NFTs on L2 liquit by preventing fragmentation.
- Mostar design sends the tokenURI to L2. We believe in continuing to use L1 as the source of truth for tokenURIs, and exploring standards such as CAIP-19 to backreference.
- There is a barrier of entry to project owners, where they have to deploy mostar. We prefer a single L1 Bridge for all NFT projects

Happy to discuss more if you disagree with my observations

**Comment by zetsuboii** [comment]
@rzmahmood Mostar used trusted actors for deployment. The holder of the NFT asks for deployment on trusted actor deploys & binds the contracts. I reasoned that NFT holders can send their tokens anywhere they want and this might just be a contract that mints a reflection on L2, and didn't include the project owner in the design. Admittedly trusted actor part was the weak side of the bridge but I was just waiting for the contract creation feature on Starknet (far as I remember it will be a 0.9 feature) to automize that

**Comment by rzmahmood** [comment]
> @rzmahmood Mostar used trusted actors for deployment. The holder of the NFT asks for deployment, trusted actor deploys & binds the contracts afterward. I reasoned that NFT holders can send their tokens anywhere they want and this might just be a contract that mints a reflection on L2, and didn't include the project owner in the design. Admittedly trusted actor part was the weak side of the bridge but I was just waiting for the contract creation feature on Starknet (far as I remember it will be a 0.9 feature) to automize that

Yep, the contract factory feature coming with 0.9 is useful, but it means that all L2 NFTs must use the same NFT contract. What if the project owner wants to have custom features in their L2 NFT?

**Comment by marcellobardus** [comment]
@rzmahmood
1. Yes with Fossil users would need to initiate an L2 txn containing the proof. The Proof can be generated using the `eth_getProof` RPC method.
2. So once a blockhash is sent to L2, let's suppose that this hash is `x` all headers in the range `(x..0)` can be retrieved this is because the `parentHash` is part of the blockheader. So one L1 message allows to claim NFTs bridged within an entire range not just a single block.
3. Yes the API makes the usage easier but is not necessary. The API essentially takes care of sending the message from L1, does what was described in point 2 if needed and then sends a webhook containing the proof once it's ready to verify.
4. The withdraw flow would be the same as in the original design.
5. The proof once verified would need to be marked as already used.
6. Maybe I was not clear enough above, in the original design the user needs to provide on L1 the L2 recipient address. In this design the user only transfers the nft to an escrow. In order to claim it on L2 we should:
* Verify the proof
* recover the signer of a message (l2_msg_sender, token_id, token_l1_address) and make sure that the signer is the EOA that sent the NFT to the L1 escrow.

**Comment by rzmahmood** [comment]
Have discussed this with @marcellobardus in a sync session.

The problems Fossil solves are awesome but introduce complexity. Our current approach is that we'll be starting off with a simple implementation and if we find that many projects are interested in bridging, then the technical overhead will be outweighed by the cost savings of the Fossil solution.

**Comment by Wave-95** [comment]
Could it be problematic if IMX registers an NFT token for L1<>L2 pair, and then the project discovers a critical issue in their L2 deployment? The project team would ask their users to withdraw their bridged assets, and instead later bridge over to a new (fixed) L2 contract at a later deployment time. 

If that scenario seems plausible, you may want to consider adding an `unregisterToken` function in this [file](https://github.com/immutable/imx-starknet/blob/v0.2.0/immutablex/ethereum/starknet/bridge/BridgeRegistry.sol) in case an L2 deployment goes wrong. 

I could see a potential use case where an old L1 contract is bridged over to a new L2 contract that houses more implementation logic (due to lower network fees existing on StarkNet). It would be great to allow the registry the flexibility to "upgrade" the L1<>L2 pairs. If for whatever reason the IMX team decided to not implement this flexibility, it may be a good idea to recommend to projects to deploy an L2 proxy contract that can later be upgraded. Just some initial thoughts when reviewing the code. Overall, great work so far.

**Comment by Wave-95** [comment]
On the withdraw side, how come it uses `permissionedBurn` instead of just sending the L2 ERC721 token into an escrow contract? Is there a reasoning for this asymmetry in terms of design? What if an L2 project only exists on StarkNet, and for whatever reason seeks to bridge into L1? (Though not entirely sure what would motivate this behavior besides seeking extra liquidity on L1 marketplaces)
